src/Titon/Kernel/AbstractKernel.hh

```diff
@@ -0,0 +1,119 @@
+<?hh // strict
+/**
+ * @copyright   2010-2015, The Titon Project
+ * @license     http://opensource.org/licenses/bsd-license.php
+ * @link        http://titon.io
+ */
+
+namespace Titon\Kernel;
+
+use Titon\Context\Depository;
+use Titon\Event\EmitsEvents;
+use Titon\Event\Subject;
+use Titon\Kernel\Middleware\Pipeline;
+
+/**
+ * The AbstractKernel is the base kernel implementation that handles the middleware pipeline process.
+ * All child kernels must implement the `handle()` method to customize the input and output handling process.
+ *
+ * @package Titon\Kernel
+ * @events
+ *      kernel.startup(Kernel $kernel, Input $input, Output $output)
+ *      kernel.shutdown(Kernel $kernel, Input $input, Output $output)
+ */
+abstract class AbstractKernel<Ti, To> extends Depository implements Kernel<Ti, To>, Subject {
+    use EmitsEvents;
+
+    /**
+     * Pipeline to manage middleware.
+     *
+     * @var \Titon\Kernel\Middleware\Pipeline
+     */
+    protected Pipeline $pipeline;
+
+    /**
+     * The time the execution started.
+     *
+     * @var float
+     */
+    protected float $startTime;
+
+    /**
+     * Instantiate a new pipeline.
+     */
+    public function __construct() {
+        $this->pipeline = new Pipeline();
```

Comment:
Yeah, just wondering how this pipeline could be customized, since it only does 2 things.

Another thing to note is that the Kernel would probably use the singleton, so that the container is available everywhere.

```hack
$app = Kernel::getInstance();
```